crypto/fipsmodule/sha/sha3_test.cc

```diff
@@ -67,6 +115,195 @@ class SHA3TestVector {
+    ASSERT_TRUE(EVP_Digest(msg_.data(), msg_.size(), digest.get(),
+                           &digest_length, algorithm, nullptr));
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    // Test Final
+    // Assert fail when |EVP_DigestFinal| is called for XOF algorithms
+    OPENSSL_memset(digest.get(), 0, digest_length);
+    ASSERT_TRUE(EVP_DigestInit(ctx.get(), algorithm));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+
+    ASSERT_FALSE(EVP_DigestFinal(ctx.get(), digest.get(), &digest_length));
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+
+    ASSERT_TRUE(EVP_DigestInit(ctx.get(), algorithm));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+    ASSERT_TRUE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+
+    // Test Final XOF after Squeeze
+    // Assert fail when |EVP_DigestFinalXOF| is called after |EVP_DigestSqueeze|
+    ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), algorithm, NULL));
+    ASSERT_TRUE(EVP_DigestSqueeze(ctx.get(), digest.get(), digest_length/2));
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get() + digest_length/2,
+                                                      digest_length/2));
+
+    // Test Update after Squeeze
+    // Assert fail when |EVP_DigestUpdate| is called after |EVP_DigestSqueeze|
+    ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), algorithm, NULL));
+    ASSERT_TRUE(EVP_DigestSqueeze(ctx.get(), digest.get(), digest_length));
+    ASSERT_FALSE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+
+    // Test Absorb
+    // Assert success when |EVP_DigestUpdate| is called byte-by-byte
+    OPENSSL_memset(digest.get(), 0, digest_length);
+    ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), algorithm, NULL));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), nullptr, 0));
+    for (const char p : msg_) {
+        ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), &p, 1));
+    }
+
+    ASSERT_TRUE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    // Test Squeeze
+    // Assert success when |EVP_DigestSqueeze| is called byte-by-byte
+    OPENSSL_memset(digest.get(), 0, digest_length);
+    ASSERT_TRUE(EVP_DigestInit(ctx.get(), algorithm));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+
+    for (size_t i = 0; i < digest_length; i++) {
+      ASSERT_TRUE(EVP_DigestSqueeze(ctx.get(), digest.get() + i, 1));
+    }
+
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    // Test Squeeze
+    // Assert success when |EVP_DigestSqueeze| is called in set byte increments
```

Comment:
Instead of hard-coding these tests in stride_tests can we just test all absorb and squeeze sizes from 1 to msg/digest length? That would also cover the above byte by byte squeeze/absord tests.

Reply:
Yeah I was thinking testing squeeze 1 byte at a time, 2 bytes at a time, 3, 4, 5, ... up to message length.

Reply:
I can add a new test doing that (I am not sure if it adds value since there is already a byte-by-byte test). However, using set increments (such as _almost_ block_size, block_size/2, etc.) tests edge-cases and I think we should keep it.